core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -26,50 +27,76 @@ import scala.annotation.tailrec
    */
   def delete(fiberRef: FiberRef[_]): FiberRefs =
-    FiberRefs(fiberRefLocals - fiberRef)
+    if (!fiberRefLocals.contains(fiberRef)) self
+    else FiberRefs(fiberRefLocals - fiberRef)
 
   /**
    * Returns a set of each `FiberRef` in this collection.
    */
   def fiberRefs: Set[FiberRef[_]] =
     fiberRefLocals.keySet
 
+  /**
+   * Boolean flag which indicates whether the FiberRefs map contains an entry
+   * that will cause the map to be changed when [[forkAs]] is called.
+   *
+   * This way we can avoid calling `fiberRefLocals.transform` on every
+   * invocation of [[forkAs]] when we already know that the map will not be
+   * transformed
+   */
+  @volatile private var needsTransformWhenForked: Boolean = true
+
   /**
    * Forks this collection of fiber refs as the specified child fiber id. This
    * will potentially modify the value of the fiber refs, as determined by the
    * individual fiber refs that make up the collection.
    */
-  def forkAs(childId: FiberId.Runtime): FiberRefs = {
-    val childFiberRefLocals: Map[FiberRef[_], (::[(FiberId.Runtime, Any, Int)], Int)] =
-      fiberRefLocals.transform { case (fiberRef, (stack, depth)) =>
-        val oldValue = stack.head._2.asInstanceOf[fiberRef.Value]
-        val newValue = fiberRef.patch(fiberRef.fork)(oldValue)
-        if (oldValue == newValue) (stack, depth) else (::((childId, newValue, 0), stack), depth + 1)
+  def forkAs(childId: FiberId.Runtime): FiberRefs =
+    if (needsTransformWhenForked) {
+      val childFiberRefLocals =
+        fiberRefLocals.transform { case (fiberRef, entry @ Value(stack, depth)) =>
+          val oldValue = stack.head.value.asInstanceOf[fiberRef.Value]
+          val newValue = fiberRef.patch(fiberRef.fork)(oldValue)
+          if (oldValue == newValue) entry
+          else Value(::(StackEntry(childId, newValue, 0), stack), depth + 1)
+        }
+      if (fiberRefLocals eq childFiberRefLocals) {
+        needsTransformWhenForked = false
+        self
+      } else {
+        FiberRefs(childFiberRefLocals)
       }
-
-    FiberRefs(childFiberRefLocals)
-  }
+    } else self
 
   /**
    * Gets the value of the specified `FiberRef` in this collection of `FiberRef`
    * values if it exists or `None` otherwise.
    */
   def get[A](fiberRef: FiberRef[A]): Option[A] =
-    fiberRefLocals.get(fiberRef).map(_._1.head._2.asInstanceOf[A])
+    Option(getOrNull(fiberRef))
 
   /**
    * Gets the value of the specified `FiberRef` in this collection of `FiberRef`
    * values if it exists or the `initial` value of the `FiberRef` otherwise.
    */
   def getOrDefault[A](fiberRef: FiberRef[A]): A =
-    get(fiberRef).getOrElse(fiberRef.initial)
+    getOrNull(fiberRef) match {
+      case null => fiberRef.initial
+      case v    => v
+    }
+
+  private[zio] def getOrNull[A](fiberRef: FiberRef[A]): A = {
+    val out = fiberRefLocals.getOrElse(fiberRef, null)
+    if (out eq null) null else out.stack.head.value
```

Comment:
```suggestion
    fiberRefLocals.get(fiberRef) match {
      case Some(out) => out.stack.head.value
      case None      => null
    }
```

avoids the allocation for the `=> default` in the `getOrElse` call

Reply:
`HashMap#getOrElse` doesn't allocate any `Option`. We benchmarked it when we did https://github.com/zio/zio/pull/8713 and `getOrElse` was much faster with less memory allocations.

Reply:
Indeed. My bad! Thanks ❤️